Failed to build with node 12.1.0

`npm install tree-sitter` failed on macOS mojave 10.14.4, with the following errors:

```
> tree-sitter@0.15.0 install /Users/zyang/node_modules/tree-sitter
> prebuild-install || node-gyp rebuild

  CC(target) Release/obj.target/tree_sitter/vendor/tree-sitter/lib/src/lib.o
  LIBTOOL-STATIC Release/tree_sitter.a
  CXX(target) Release/obj.target/tree_sitter_runtime_binding/src/binding.o
  CXX(target) Release/obj.target/tree_sitter_runtime_binding/src/conversions.o
```
```
../src/conversions.cc:105:41: error: too few arguments to function call, single
      argument 'context' was not specified
  if (std::isfinite(js_row->NumberValue())) {
                    ~~~~~~~~~~~~~~~~~~~ ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8.h:2564:3: note: 'NumberValue'
      declared here
  V8_WARN_UNUSED_RESULT Maybe<double> NumberValue(Local<Context> context) const;
  ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded
      from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:106:52: error: too few arguments to function call, single
      argument 'context' was not specified
    row = static_cast<uint32_t>(js_row->Int32Value());
                                ~~~~~~~~~~~~~~~~~~ ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8.h:2569:3: note: 'Int32Value'
      declared here
  V8_WARN_UNUSED_RESULT Maybe<int32_t> Int32Value(Local<Context> context) const;
  ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded
      from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:111:44: error: too few arguments to function call, single
      argument 'context' was not specified
  if (std::isfinite(js_column->NumberValue())) {
                    ~~~~~~~~~~~~~~~~~~~~~~ ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8.h:2564:3: note: 'NumberValue'
      declared here
  V8_WARN_UNUSED_RESULT Maybe<double> NumberValue(Local<Context> context) const;
  ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded
      from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:112:58: error: too few arguments to function call, single
      argument 'context' was not specified
    column = static_cast<uint32_t>(js_column->Int32Value()) * BYTES_PER_...
                                   ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8.h:2569:3: note: 'Int32Value'
      declared here
  V8_WARN_UNUSED_RESULT Maybe<int32_t> Int32Value(Local<Context> context) const;
  ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded
      from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:130:47: error: too few arguments to function call, single
      argument 'context' was not specified
  return Nan::Just<uint32_t>(arg->Uint32Value() * BYTES_PER_CHARACTER);
                             ~~~~~~~~~~~~~~~~ ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8.h:2567:3: note: 'Uint32Value'
      declared here
  V8_WARN_UNUSED_RESULT Maybe<uint32_t> Uint32Value(
  ^
/Users/zyang/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded
      from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
13 warnings and 5 errors generated.
make: *** [Release/obj.target/tree_sitter_runtime_binding/src/conversions.o] Error 1
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:256:12)
gyp ERR! System Darwin 18.5.0
gyp ERR! command "/usr/local/Cellar/node/12.1.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/zyang/node_modules/tree-sitter
gyp ERR! node -v v12.1.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
```

It seems that `conversions.cc` uses in-capatible v8 API?

rebuild-install WARN install No prebuilt binaries found (target=12.1.0 runtime=node arch=x64 libc= platform=darwin)
  CC(target) Release/obj.target/tree_sitter/vendor/tree-sitter/lib/src/lib.o
  LIBTOOL-STATIC Release/tree_sitter.a
  CXX(target) Release/obj.target/tree_sitter_runtime_binding/src/binding.o
  CXX(target) Release/obj.target/tree_sitter_runtime_binding/src/conversions.o
../src/conversions.cc:32:12: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  exports->Set(Nan::New("pointTransferArray").ToLocalChecked(), Uint32Array::New(js_point_transfer_buffer, 0, 2));
           ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:42:11: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  result->Set(Nan::New(start_position_key), PointToJS(range.start_point));
          ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:43:11: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  result->Set(Nan::New(start_index_key), ByteCountToJS(range.start_byte));
   I'm also facing similar issue on OSX Mojave 10.14.5

/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:44:11: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  result->Set(Nan::New(end_position_key), PointToJS(range.end_point));
          ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:45:11: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  result->Set(Nan::New(end_index_key), ByteCountToJS(range.end_byte));
          ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:68:3: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  INIT(start_point, start_position_key, PointFromJS);
  ^
../src/conversions.cc:60:33: note: expanded from macro 'INIT'
    auto field = Type(js_range->Get(Nan::New(key))); \
                                ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:69:3: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  INIT(end_point, end_position_key, PointFromJS);
  ^
../src/conversions.cc:60:33: note: expanded from macro 'INIT'
    auto field = Type(js_range->Get(Nan::New(key))); \
                                ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:70:3: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  INIT(start_byte, start_index_key, ByteCountFromJS);
  ^
../src/conversions.cc:60:33: note: expanded from macro 'INIT'
    auto field = Type(js_range->Get(Nan::New(key))); \
                                ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:71:3: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  INIT(end_byte, end_index_key, ByteCountFromJS);
  ^
../src/conversions.cc:60:33: note: expanded from macro 'INIT'
    auto field = Type(js_range->Get(Nan::New(key))); \
                                ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:80:11: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  result->Set(Nan::New(row_key), Nan::New<Number>(point.row));
          ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:81:11: warning: 'Set' is deprecated: Use maybe version [-Wdeprecated-declarations]
  result->Set(Nan::New(column_key), ByteCountToJS(point.column));
          ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3358:3: note: 'Set' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version",
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:92:35: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  Local<Value> js_row = js_point->Get(Nan::New(row_key));
                                  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:98:38: warning: 'Get' is deprecated: Use maybe version [-Wdeprecated-declarations]
  Local<Value> js_column = js_point->Get(Nan::New(column_key));
                                     ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:3412:3: note: 'Get' has been explicitly marked deprecated here
  V8_DEPRECATE_SOON("Use maybe version", Local<Value> Get(Local<Value> key));
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:322:29: note: expanded from macro 'V8_DEPRECATE_SOON'
  declarator __attribute__((deprecated(message)))
                            ^
../src/conversions.cc:105:41: error: too few arguments to function call, single argument 'context' was not specified
  if (std::isfinite(js_row->NumberValue())) {
                    ~~~~~~~~~~~~~~~~~~~ ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:2564:3: note: 'NumberValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<double> NumberValue(Local<Context> context) const;
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:106:52: error: too few arguments to function call, single argument 'context' was not specified
    row = static_cast<uint32_t>(js_row->Int32Value());
                                ~~~~~~~~~~~~~~~~~~ ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:2569:3: note: 'Int32Value' declared here
  V8_WARN_UNUSED_RESULT Maybe<int32_t> Int32Value(Local<Context> context) const;
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:111:44: error: too few arguments to function call, single argument 'context' was not specified
  if (std::isfinite(js_column->NumberValue())) {
                    ~~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:2564:3: note: 'NumberValue' declared here
  V8_WARN_UNUSED_RESULT Maybe<double> NumberValue(Local<Context> context) const;
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:112:58: error: too few arguments to function call, single argument 'context' was not specified
    column = static_cast<uint32_t>(js_column->Int32Value()) * BYTES_PER_CHARACTER;
                                   ~~~~~~~~~~~~~~~~~~~~~ ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:2569:3: note: 'Int32Value' declared here
  V8_WARN_UNUSED_RESULT Maybe<int32_t> Int32Value(Local<Context> context) const;
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../src/conversions.cc:130:47: error: too few arguments to function call, single argument 'context' was not specified
  return Nan::Just<uint32_t>(arg->Uint32Value() * BYTES_PER_CHARACTER);
                             ~~~~~~~~~~~~~~~~ ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8.h:2567:3: note: 'Uint32Value' declared here
  V8_WARN_UNUSED_RESULT Maybe<uint32_t> Uint32Value(
  ^
/Users/mayankmittal/.node-gyp/12.1.0/include/node/v8config.h:347:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))


Is there workaround to go through it?

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,6 +2,7 @@ language: node_js
 
 node_js:
 - 10
+- 12
 
 matrix:
   include:

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,18 +4,9 @@ node_js:
 - 10
 - 12
 
-matrix:
-  include:
-  - os: osx
-  - os: linux
-    env:
-      - CXX=clang++
-      - CC=clang
-    addons:
-      apt:
-        packages:
-        - build-essential
-        - clang-3.3
+os:
+- osx
+- linux
 
 before_install:
 - git submodule update --init --recursive

--- a/package.json
+++ b/package.json
@@ -19,10 +19,10 @@
     "prebuild-install": "^5.0.0"
   },
   "devDependencies": {
-    "prebuild": "^7.6.0",
     "chai": "3.5.x",
     "mocha": "^5.2.0",
-    "superstring": "^2.3.5",
+    "prebuild": "^7.6.0",
+    "superstring": "^2.4.1",
     "tree-sitter-javascript": "git://github.com/tree-sitter/tree-sitter-javascript.git#master"
   },
   "scripts": {

--- a/src/conversions.cc
+++ b/src/conversions.cc
@@ -29,7 +29,7 @@ void InitConversions(Local<Object> exports) {
 
   point_transfer_buffer = static_cast<uint32_t *>(malloc(2 * sizeof(uint32_t)));
   auto js_point_transfer_buffer = ArrayBuffer::New(Isolate::GetCurrent(), point_transfer_buffer, 2 * sizeof(uint32_t));
-  exports->Set(Nan::New("pointTransferArray").ToLocalChecked(), Uint32Array::New(js_point_transfer_buffer, 0, 2));
+  Nan::Set(exports, Nan::New("pointTransferArray").ToLocalChecked(), Uint32Array::New(js_point_transfer_buffer, 0, 2));
 }
 
 void TransferPoint(const TSPoint &point) {
@@ -39,10 +39,10 @@ void TransferPoint(const TSPoint &point) {
 
 Local<Object> RangeToJS(const TSRange &range) {
   Local<Object> result = Nan::New<Object>();
-  result->Set(Nan::New(start_position_key), PointToJS(range.start_point));
-  result->Set(Nan::New(start_index_key), ByteCountToJS(range.start_byte));
-  result->Set(Nan::New(end_position_key), PointToJS(range.end_point));
-  result->Set(Nan::New(end_index_key), ByteCountToJS(range.end_byte));
+  Nan::Set(result, Nan::New(start_position_key), PointToJS(range.start_point));
+  Nan::Set(result, Nan::New(start_index_key), ByteCountToJS(range.start_byte));
+  Nan::Set(result, Nan::New(end_position_key), PointToJS(range.end_point));
+  Nan::Set(result, Nan::New(end_index_key), ByteCountToJS(range.end_byte));
   return result;
 }
 
@@ -56,8 +56,13 @@ Nan::Maybe<TSRange> RangeFromJS(const Local<Value> &arg) {
 
   Local<Object> js_range = Local<Object>::Cast(arg);
 
-  #define INIT(field, key, Type) { \
-    auto field = Type(js_range->Get(Nan::New(key))); \
+  #define INIT(field, key, Convert) { \
+    auto value = Nan::Get(js_range, Nan::New(key)); \
+    if (value.IsEmpty()) { \
+      Nan::ThrowTypeError("Range must be a {startPosition, endPosition, startIndex, endIndex} object"); \
+      return Nan::Nothing<TSRange>(); \
+    } \
+    auto field = Convert(value.ToLocalChecked()); \
     if (field.IsJust()) { \
       result.field = field.FromJust(); \
     } else { \
@@ -77,41 +82,48 @@ Nan::Maybe<TSRange> RangeFromJS(const Local<Value> &arg) {
 
 Local<Object> PointToJS(const TSPoint &point) {
   Local<Object> result = Nan::New<Object>();
-  result->Set(Nan::New(row_key), Nan::New<Number>(point.row));
-  result->Set(Nan::New(column_key), ByteCountToJS(point.column));
+  Nan::Set(result, Nan::New(row_key), Nan::New<Number>(point.row));
+  Nan::Set(result, Nan::New(column_key), ByteCountToJS(point.column));
   return result;
 }
 
 Nan::Maybe<TSPoint> PointFromJS(const Local<Value> &arg) {
-  if (!arg->IsObject()) {
+  Local<Object> js_point;
+  if (!arg->IsObject() || !Nan::To<Object>(arg).ToLocal(&js_point)) {
     Nan::ThrowTypeError("Point must be a {row, column} object");
     return Nan::Nothing<TSPoint>();
   }
 
-  Local<Object> js_point = Local<Object>::Cast(arg);
-  Local<Value> js_row = js_point->Get(Nan::New(row_key));
-  if (!js_row->IsNumber()) {
+  Local<Value> js_row;
+  if (!Nan::Get(js_point, Nan::New(row_key)).ToLocal(&js_row)) {
+    Nan::ThrowTypeError("Point must be a {row, column} object");
+    return Nan::Nothing<TSPoint>();
+  }
+
+  Local<Value> js_column;
+  if (!Nan::Get(js_point, Nan::New(column_key)).ToLocal(&js_column)) {
+    Nan::ThrowTypeError("Point must be a {row, column} object");
+    return Nan::Nothing<TSPoint>();
+  }
+
+  uint32_t row;
+  if (!std::isfinite(Nan::To<double>(js_row).FromMaybe(0))) {
+    row = UINT32_MAX;
+  } else if (js_row->IsNumber()) {
+    row = Nan::To<uint32_t>(js_row).FromJust();
+  } else {
     Nan::ThrowTypeError("Point.row must be a number");
     return Nan::Nothing<TSPoint>();
   }
 
-  Local<Value> js_column = js_point->Get(Nan::New(column_key));
-  if (!js_column->IsNumber()) {
+  uint32_t column;
+  if (!std::isfinite(Nan::To<double>(js_column).FromMaybe(0))) {
+    column = UINT32_MAX;
+  } else if (js_column->IsNumber()) {
+    column = Nan::To<uint32_t>(js_column).FromMaybe(0) * BYTES_PER_CHARACTER;
+  } else {
     Nan::ThrowTypeError("Point.column must be a number");
     return Nan::Nothing<TSPoint>();
-  }
-
-  uint32_t row, column;
-  if (std::isfinite(js_row->NumberValue())) {
-    row = static_cast<uint32_t>(js_row->Int32Value());
-  } else {
-    row = UINT32_MAX;
-  }
-
-  if (std::isfinite(js_column->NumberValue())) {
-    column = static_cast<uint32_t>(js_column->Int32Value()) * BYTES_PER_CHARACTER;
-  } else {
-    column = UINT32_MAX;
   }
 
   return Nan::Just<TSPoint>({row, column});
@@ -122,12 +134,13 @@ Local<Number> ByteCountToJS(uint32_t byte_count) {
 }
 
 Nan::Maybe<uint32_t> ByteCountFromJS(const v8::Local<v8::Value> &arg) {
-  if (!arg->IsUint32()) {
+  auto result = Nan::To<uint32_t>(arg);
+  if (!arg->IsNumber()) {
     Nan::ThrowTypeError("Character index must be a number");
     return Nan::Nothing<uint32_t>();
   }
 
-  return Nan::Just<uint32_t>(arg->Uint32Value() * BYTES_PER_CHARACTER);
+  return Nan::Just<uint32_t>(result.FromJust() * BYTES_PER_CHARACTER);
 }
 
 }  // namespace node_tree_sitter

--- a/src/language.cc
+++ b/src/language.cc
@@ -48,9 +48,9 @@ static void GetNodeTypeNamesById(const Nan::FunctionCallbackInfo<Value> &info) {
     const char *name = ts_language_symbol_name(language, i);
     TSSymbolType type = ts_language_symbol_type(language, i);
     if (type == TSSymbolTypeRegular) {
-      result->Set(i, Nan::New(name).ToLocalChecked());
+      Nan::Set(result, i, Nan::New(name).ToLocalChecked());
     } else {
-      result->Set(i, Nan::Null());
+      Nan::Set(result, i, Nan::Null());
     }
   }
 
@@ -66,23 +66,25 @@ static void GetNodeFieldNamesById(const Nan::FunctionCallbackInfo<Value> &info) 
   for (uint32_t i = 0; i < length + 1; i++) {
     const char *name = ts_language_field_name_for_id(language, i);
     if (name) {
-      result->Set(i, Nan::New(name).ToLocalChecked());
+      Nan::Set(result, i, Nan::New(name).ToLocalChecked());
     } else {
-      result->Set(i, Nan::Null());
+      Nan::Set(result, i, Nan::Null());
     }
   }
   info.GetReturnValue().Set(result);
 }
 
 void Init(Local<Object> exports) {
-  exports->Set(
+  Nan::Set(
+    exports,
     Nan::New("getNodeTypeNamesById").ToLocalChecked(),
-    Nan::New<FunctionTemplate>(GetNodeTypeNamesById)->GetFunction()
+    Nan::GetFunction(Nan::New<FunctionTemplate>(GetNodeTypeNamesById)).ToLocalChecked()
   );
 
-  exports->Set(
+  Nan::Set(
+    exports,
     Nan::New("getNodeFieldNamesById").ToLocalChecked(),
-    Nan::New<FunctionTemplate>(GetNodeFieldNamesById)->GetFunction()
+    Nan::GetFunction(Nan::New<FunctionTemplate>(GetNodeFieldNamesById)).ToLocalChecked()
   );
 }
 

--- a/src/logger.cc
+++ b/src/logger.cc
@@ -36,21 +36,21 @@ void Logger::Log(void *payload, TSLogType type, const char *message_str) {
 
     string key = message.substr(key_pos, (value_sep_pos - key_pos));
     string value = message.substr(val_pos, (param_sep_pos - val_pos));
-    params->Set(Nan::New(key).ToLocalChecked(), Nan::New(value).ToLocalChecked());
+    Nan::Set(params, Nan::New(key).ToLocalChecked(), Nan::New(value).ToLocalChecked());
   }
 
   Local<Value> argv[3] = { name, params, type_name };
   TryCatch try_catch(Isolate::GetCurrent());
-  fn->Call(fn->CreationContext()->Global(), 3, argv);
+  Nan::Call(fn, fn->CreationContext()->Global(), 3, argv);
   if (try_catch.HasCaught()) {
     Local<Value> log_argv[2] = {
       Nan::New("Error in debug callback:").ToLocalChecked(),
       try_catch.Exception()
     };
 
-    Local<Object> console = Local<Object>::Cast(fn->CreationContext()->Global()->Get(Nan::New("console").ToLocalChecked()));
-    Local<Function> error_fn = Local<Function>::Cast(console->Get(Nan::New("error").ToLocalChecked()));
-    error_fn->Call(console, 2, log_argv);
+    Local<Object> console = Local<Object>::Cast(Nan::Get(fn->CreationContext()->Global(), Nan::New("console").ToLocalChecked()).ToLocalChecked());
+    Local<Function> error_fn = Local<Function>::Cast(Nan::Get(console, Nan::New("error").ToLocalChecked()).ToLocalChecked());
+    Nan::Call(error_fn, console, 2, log_argv);
   }
 }
 

--- a/src/node.cc
+++ b/src/node.cc
@@ -27,7 +27,8 @@ static inline void setup_transfer_buffer(uint32_t node_count) {
     transfer_buffer_length = new_length;
     transfer_buffer = static_cast<uint32_t *>(malloc(transfer_buffer_length * sizeof(uint32_t)));
     auto js_transfer_buffer = ArrayBuffer::New(Isolate::GetCurrent(), transfer_buffer, transfer_buffer_length * sizeof(uint32_t));
-    Nan::New(module_exports)->Set(
+    Nan::Set(
+      Nan::New(module_exports),
       Nan::New("nodeTransferArray").ToLocalChecked(),
       Uint32Array::New(js_transfer_buffer, 0, transfer_buffer_length)
     );
@@ -56,13 +57,12 @@ static void MarshalNodes(const Nan::FunctionCallbackInfo<Value> &info,
       *(p++) = node.context[2];
       *(p++) = node.context[3];
       if (node.id) {
-        result->Set(i, Nan::New(ts_node_symbol(node)));
+        Nan::Set(result, i, Nan::New(ts_node_symbol(node)));
       } else {
-        result->Set(i, Nan::Null());
+        Nan::Set(result, i, Nan::Null());
       }
     } else {
-      assert(!cache_entry->second->node.IsNearDeath());
-      result->Set(i, Nan::New(cache_entry->second->node));
+      Nan::Set(result, i, Nan::New(cache_entry->second->node));
     }
   }
   info.GetReturnValue().Set(result);
@@ -83,7 +83,6 @@ void MarshalNode(const Nan::FunctionCallbackInfo<Value> &info, const Tree *tree,
       info.GetReturnValue().Set(Nan::New(ts_node_symbol(node)));
     }
   } else {
-    assert(!cache_entry->second->node.IsNearDeath());
     info.GetReturnValue().Set(Nan::New(cache_entry->second->node));
   }
 }
@@ -317,7 +316,7 @@ static void Child(const Nan::FunctionCallbackInfo<Value> &info) {
       Nan::ThrowTypeError("Second argument must be an integer");
       return;
     }
-    uint32_t index = info[1]->Uint32Value();
+    uint32_t index = Nan::To<uint32_t>(info[1]).FromJust();
     MarshalNode(info, tree, ts_node_child(node, index));
     return;
   }
@@ -333,7 +332,7 @@ static void NamedChild(const Nan::FunctionCallbackInfo<Value> &info) {
       Nan::ThrowTypeError("Second argument must be an integer");
       return;
     }
-    uint32_t index = info[1]->Uint32Value();
+    uint32_t index = Nan::To<uint32_t>(info[1]).FromJust();
     MarshalNode(info, tree, ts_node_named_child(node, index));
     return;
   }
@@ -470,26 +469,37 @@ bool symbol_set_from_js(SymbolSet *symbols, const Local<Value> &value, const TSL
 
   Local<Array> js_types = Local<Array>::Cast(value);
   for (unsigned i = 0, n = js_types->Length(); i < n; i++) {
-    Local<Value> js_types_i = js_types->Get(i);
-    if (!js_types_i->IsString()) {
-      Nan::ThrowTypeError("Argument must be a string or array of strings");
-      return false;
-    }
+    Local<Value> js_node_type_value;
+    if (Nan::Get(js_types, i).ToLocal(&js_node_type_value)) {
+      Local<String> js_node_type;
+      if (Nan::To<String>(js_node_type_value).ToLocal(&js_node_type)) {
+        std::string node_type(js_node_type->Utf8Length(Isolate::GetCurrent()), '\0');
+        js_node_type->WriteUtf8(
 
-    Local<String> js_node_type = Local<String>::Cast(js_types_i);
-    std::string node_type(js_node_type->Utf8Length(), '\0');
-    js_node_type->WriteUtf8(&node_type[0]);
+          // Nan doesn't wrap this functionality
+          #if NODE_MAJOR_VERSION >= 12
+                Isolate::GetCurrent(),
+          #endif
 
-    if (node_type == "ERROR") {
-      symbols->add(static_cast<TSSymbol>(-1));
-      continue;
-    }
+          &node_type[0]
+        );
 
-    for (TSSymbol j = 0; j < symbol_count; j++) {
-      if (node_type == ts_language_symbol_name(language, j)) {
-        symbols->add(j);
+        if (node_type == "ERROR") {
+          symbols->add(static_cast<TSSymbol>(-1));
+        } else {
+          for (TSSymbol j = 0; j < symbol_count; j++) {
+            if (node_type == ts_language_symbol_name(language, j)) {
+              symbols->add(j);
+            }
+          }
+        }
+
+        continue;
       }
     }
+
+    Nan::ThrowTypeError("Argument must be a string or array of strings");
+    return false;
   }
 
   return true;
@@ -542,13 +552,13 @@ static void DescendantsOfType(const Nan::FunctionCallbackInfo<Value> &info) {
   TSPoint start_point = {0, 0};
   TSPoint end_point = {UINT32_MAX, UINT32_MAX};
 
-  if (info[2]->BooleanValue()) {
+  if (info.Length() > 2 && info[2]->IsObject()) {
     auto maybe_start_point = PointFromJS(info[2]);
     if (maybe_start_point.IsNothing()) return;
     start_point = maybe_start_point.FromJust();
   }
 
-  if (info[3]->BooleanValue()) {
+  if (info.Length() > 3 && info[3]->IsObject()) {
     auto maybe_end_point = PointFromJS(info[3]);
     if (maybe_end_point.IsNothing()) return;
     end_point = maybe_end_point.FromJust();
@@ -602,11 +612,12 @@ static void ChildNodesForFieldId(const Nan::FunctionCallbackInfo<Value> &info) {
   TSNode node = UnmarshalNode(tree);
   if (!node.id) return;
 
-  if (!info[1]->IsUint32()) {
+  auto maybe_field_id = Nan::To<uint32_t>(info[1]);
+  if (!maybe_field_id.IsJust()) {
     Nan::ThrowTypeError("Second argument must be an integer");
     return;
   }
-  uint32_t field_id = info[1]->Uint32Value();
+  uint32_t field_id = maybe_field_id.FromJust();
 
   vector<TSNode> result;
   ts_tree_cursor_reset(&scratch_cursor, node);
@@ -627,11 +638,12 @@ static void ChildNodeForFieldId(const Nan::FunctionCallbackInfo<Value> &info) {
   TSNode node = UnmarshalNode(tree);
 
   if (node.id) {
-    if (!info[1]->IsUint32()) {
+    auto maybe_field_id = Nan::To<uint32_t>(info[1]);
+    if (!maybe_field_id.IsJust()) {
       Nan::ThrowTypeError("Second argument must be an integer");
       return;
     }
-    uint32_t field_id = info[1]->Uint32Value();
+    uint32_t field_id = maybe_field_id.FromJust();
     MarshalNode(info, tree, ts_node_child_by_field_id(node, field_id));
     return;
   }
@@ -710,16 +722,17 @@ void Init(Local<Object> exports) {
   };
 
   for (size_t i = 0; i < length_of_array(methods); i++) {
-    result->Set(
+    Nan::Set(
+      result,
       Nan::New(methods[i].name).ToLocalChecked(),
-      Nan::New<FunctionTemplate>(methods[i].callback)->GetFunction()
+      Nan::GetFunction(Nan::New<FunctionTemplate>(methods[i].callback)).ToLocalChecked()
     );
   }
 
   module_exports.Reset(exports);
   setup_transfer_buffer(1);
 
-  exports->Set(Nan::New("NodeMethods").ToLocalChecked(), result);
+  Nan::Set(exports, Nan::New("NodeMethods").ToLocalChecked(), result);
 }
 
 }  // namespace node_methods

--- a/src/parser.cc
+++ b/src/parser.cc
@@ -26,11 +26,9 @@ class CallbackInput {
     : callback(callback),
       byte_offset(0),
       partial_string_offset(0) {
-    if (js_buffer_size->IsUint32()) {
-      buffer.resize(Local<Integer>::Cast(js_buffer_size)->Uint32Value());
-    } else {
-      buffer.resize(32 * 1024);
-    }
+    uint32_t buffer_size = Nan::To<uint32_t>(js_buffer_size).FromMaybe(0);
+    if (buffer_size == 0) buffer_size = 32 * 1024;
+    buffer.resize(buffer_size);
   }
 
   TSInput Input() {
@@ -51,8 +49,9 @@ class CallbackInput {
       reader->partial_string.Reset();
     }
 
+    *bytes_read = 0;
     Local<String> result;
-    uint32_t start;
+    uint32_t start = 0;
     if (reader->partial_string_offset) {
       result = Nan::New(reader->partial_string);
       start = reader->partial_string_offset;
@@ -61,18 +60,27 @@ class CallbackInput {
       uint32_t utf16_unit = byte / 2;
       Local<Value> argv[2] = { Nan::New<Number>(utf16_unit), PointToJS(position) };
       TryCatch try_catch(Isolate::GetCurrent());
-      auto result_value = callback->Call(Nan::Null(), 2, argv);
-      if (!try_catch.HasCaught() && result_value->IsString()) {
-        result = Local<String>::Cast(result_value);
-        start = 0;
-      } else {
-        *bytes_read = 0;
-        start = 0;
-        return "";
-      }
+      auto maybe_result_value = Nan::Call(callback, callback->CreationContext()->Global(), 2, argv);
+      if (try_catch.HasCaught()) return nullptr;
+
+      Local<Value> result_value;
+      if (!maybe_result_value.ToLocal(&result_value)) return nullptr;
+      if (!result_value->IsString()) return nullptr;
+      if (!Nan::To<String>(result_value).ToLocal(&result)) return nullptr;
     }
 
-    int utf16_units_read = result->Write(reader->buffer.data(), start, reader->buffer.size(), 2);
+    int utf16_units_read = result->Write(
+
+      // Nan doesn't wrap this functionality
+      #if NODE_MAJOR_VERSION >= 12
+            Isolate::GetCurrent(),
+      #endif
+
+      reader->buffer.data(),
+      start,
+      reader->buffer.size(),
+      String::NO_NULL_TERMINATION
+    );
     int end = start + utf16_units_read;
     *bytes_read = 2 * utf16_units_read;
 
@@ -173,9 +181,9 @@ void Parser::Init(Local<Object> exports) {
     Nan::SetPrototypeMethod(tpl, methods[i].name, methods[i].callback);
   }
 
-  constructor.Reset(Nan::Persistent<Function>(tpl->GetFunction()));
-  exports->Set(class_name, Nan::New(constructor));
-  exports->Set(Nan::New("LANGUAGE_VERSION").ToLocalChecked(), Nan::New<Number>(TREE_SITTER_LANGUAGE_VERSION));
+  constructor.Reset(Nan::Persistent<Function>(Nan::GetFunction(tpl).ToLocalChecked()));
+  Nan::Set(exports, class_name, Nan::New(constructor));
+  Nan::Set(exports, Nan::New("LANGUAGE_VERSION").ToLocalChecked(), Nan::New<Number>(TREE_SITTER_LANGUAGE_VERSION));
 }
 
 Parser::Parser() : parser_(ts_parser_new()), is_parsing_async_(false) {}
@@ -188,7 +196,9 @@ static bool handle_included_ranges(TSParser *parser, Local<Value> arg) {
     auto js_included_ranges = Local<Array>::Cast(arg);
     vector<TSRange> included_ranges;
     for (unsigned i = 0; i < js_included_ranges->Length(); i++) {
-      auto maybe_range = RangeFromJS(js_included_ranges->Get(i));
+      Local<Value> range_value;
+      if (!Nan::Get(js_included_ranges, i).ToLocal(&range_value)) return false;
+      auto maybe_range = RangeFromJS(range_value);
       if (!maybe_range.IsJust()) return false;
       auto range = maybe_range.FromJust();
       if (range.start_byte < last_included_range_end) {
@@ -250,9 +260,10 @@ void Parser::Parse(const Nan::FunctionCallbackInfo<Value> &info) {
 
   Local<Function> callback = Local<Function>::Cast(info[0]);
 
+  Local<Object> js_old_tree;
   const TSTree *old_tree = nullptr;
-  if (info.Length() > 1 && info[1]->BooleanValue()) {
-    const Tree *tree = Tree::UnwrapTree(info[1]);
+  if (info.Length() > 1 && !info[1]->IsNull() && !info[1]->IsUndefined() && Nan::To<Object>(info[1]).ToLocal(&js_old_tree)) {
+    const Tree *tree = Tree::UnwrapTree(js_old_tree);
     if (!tree) {
       Nan::ThrowTypeError("Second argument must be a tree");
       return;
@@ -305,9 +316,10 @@ void Parser::ParseTextBuffer(const Nan::FunctionCallbackInfo<Value> &info) {
     return;
   }
 
+  Local<Object> js_old_tree;
   const TSTree *old_tree = nullptr;
-  if (info.Length() > 2 && info[2]->BooleanValue()) {
-    const Tree *tree = Tree::UnwrapTree(info[2]);
+  if (info.Length() > 2 && info[2]->IsObject() && Nan::To<Object>(info[2]).ToLocal(&js_old_tree)) {
+    const Tree *tree = Tree::UnwrapTree(js_old_tree);
     if (!tree) {
       Nan::ThrowTypeError("Second argument must be a tree");
       return;
@@ -322,18 +334,21 @@ void Parser::ParseTextBuffer(const Nan::FunctionCallbackInfo<Value> &info) {
 
   // If a `syncTimeoutMicros` option is passed, parse synchronously
   // for the given amount of time before queuing an async task.
-  if (info[4]->BooleanValue()) {
-    double js_sync_timeout = info[4]->NumberValue();
+  double js_sync_timeout = Nan::To<double>(info[4]).FromMaybe(-1);
+  if (js_sync_timeout > 0) {
     size_t sync_timeout;
 
     // If the timeout is `Infinity`, then parse synchronously with no timeout.
-    if (js_sync_timeout > 0 && !std::isfinite(js_sync_timeout)) {
+    if (js_sync_timeout && !std::isfinite(js_sync_timeout)) {
       sync_timeout = 0;
-    } else if (info[4]->IsUint32()) {
-      sync_timeout = info[4]->Uint32Value();
     } else {
-      Nan::ThrowTypeError("The `syncTimeoutMicros` option must be a positive integer.");
-      return;
+      auto maybe_sync_timeout = Nan::To<uint32_t>(info[4]);
+      if (maybe_sync_timeout.IsJust()) {
+        sync_timeout = maybe_sync_timeout.FromJust();
+      } else {
+        Nan::ThrowTypeError("The `syncTimeoutMicros` option must be a positive integer.");
+        return;
+      }
     }
 
     // Logging is disabled for this method, because we can't call the
@@ -348,7 +363,8 @@ void Parser::ParseTextBuffer(const Nan::FunctionCallbackInfo<Value> &info) {
     if (result) {
       delete input;
       Local<Value> argv[] = {Tree::NewInstance(result)};
-      info[0].As<Function>()->Call(Nan::Null(), 1, argv);
+      auto callback = info[0].As<Function>();
+      Nan::Call(callback, callback->CreationContext()->Global(), 1, argv);
       return;
     }
   }
@@ -369,9 +385,10 @@ void Parser::ParseTextBufferSync(const Nan::FunctionCallbackInfo<Value> &info) {
     return;
   }
 
-  TSTree *old_tree = nullptr;
-  if (info.Length() > 1 && info[1]->BooleanValue()) {
-    const Tree *tree = Tree::UnwrapTree(info[1]);
+  Local<Object> js_old_tree;
+  const TSTree *old_tree = nullptr;
+  if (info.Length() > 1 && info[1]->IsObject() && Nan::To<Object>(info[1]).ToLocal(&js_old_tree)) {
+    const Tree *tree = Tree::UnwrapTree(js_old_tree);
     if (!tree) {
       Nan::ThrowTypeError("Second argument must be a tree");
       return;
@@ -411,7 +428,7 @@ void Parser::SetLogger(const Nan::FunctionCallbackInfo<Value> &info) {
   if (info[0]->IsFunction()) {
     if (current_logger.payload) delete (Logger *)current_logger.payload;
     ts_parser_set_logger(parser->parser_, Logger::Make(Local<Function>::Cast(info[0])));
-  } else if (!info[0]->BooleanValue()) {
+  } else if (!Nan::To<bool>(info[0]).FromMaybe(true)) {
     if (current_logger.payload) delete (Logger *)current_logger.payload;
     ts_parser_set_logger(parser->parser_, { 0, 0 });
   } else {
@@ -429,9 +446,7 @@ void Parser::PrintDotGraphs(const Nan::FunctionCallbackInfo<Value> &info) {
     return;
   }
 
-  Local<Boolean> value = Local<Boolean>::Cast(info[0]);
-
-  if (value->IsBoolean() && value->BooleanValue()) {
+  if (Nan::To<bool>(info[0]).FromMaybe(false)) {
     ts_parser_print_dot_graphs(parser->parser_, 2);
   } else {
     ts_parser_print_dot_graphs(parser->parser_, -1);

--- a/src/tree.cc
+++ b/src/tree.cc
@@ -34,11 +34,11 @@ void Tree::Init(Local<Object> exports) {
     Nan::SetPrototypeMethod(tpl, methods[i].name, methods[i].callback);
   }
 
-  Local<Function> ctor = tpl->GetFunction();
+  Local<Function> ctor = Nan::GetFunction(tpl).ToLocalChecked();
 
   constructor_template.Reset(tpl);
   constructor.Reset(ctor);
-  exports->Set(class_name, ctor);
+  Nan::Set(exports, class_name, ctor);
 }
 
 Tree::Tree(TSTree *tree) : tree_(tree) {}
@@ -53,7 +53,7 @@ Tree::~Tree() {
 Local<Value> Tree::NewInstance(TSTree *tree) {
   if (tree) {
     Local<Object> self;
-    MaybeLocal<Object> maybe_self = Nan::New(constructor)->NewInstance(Nan::GetCurrentContext());
+    MaybeLocal<Object> maybe_self = Nan::NewInstance(Nan::New(constructor));
     if (maybe_self.ToLocal(&self)) {
       (new Tree(tree))->Wrap(self);
       return self;
@@ -69,15 +69,15 @@ const Tree *Tree::UnwrapTree(const Local<Value> &value) {
   return ObjectWrap::Unwrap<Tree>(js_tree);
 }
 
-void Tree::New(const Nan::FunctionCallbackInfo<Value> &info) {
-  info.GetReturnValue().Set(Nan::Null());
-}
+void Tree::New(const Nan::FunctionCallbackInfo<Value> &info) {}
 
 #define read_number_from_js(out, value, name)        \
-  if (!(value)->IsUint32()) {                       \
+  maybe_number = Nan::To<uint32_t>(value);           \
+  if (maybe_number.IsNothing()) {                    \
     Nan::ThrowTypeError(name " must be an integer"); \
+    return;                                          \
   }                                                  \
-  *(out) = (value)->Uint32Value()
+  *(out) = maybe_number.FromJust();
 
 #define read_byte_count_from_js(out, value, name)   \
   read_number_from_js(out, value, name);            \
@@ -87,6 +87,7 @@ void Tree::Edit(const Nan::FunctionCallbackInfo<Value> &info) {
   Tree *tree = ObjectWrap::Unwrap<Tree>(info.This());
 
   TSInputEdit edit;
+  Nan::Maybe<uint32_t> maybe_number = Nan::Nothing<uint32_t>();
   read_number_from_js(&edit.start_point.row, info[0], "startPosition.row");
   read_byte_count_from_js(&edit.start_point.column, info[1], "startPosition.column");
   read_number_from_js(&edit.old_end_point.row, info[2], "oldEndPosition.row");
@@ -104,13 +105,16 @@ void Tree::Edit(const Nan::FunctionCallbackInfo<Value> &info) {
     TSNode node;
     node.id = entry.first;
     for (unsigned i = 0; i < 4; i++) {
-      node.context[i] = js_node->Get(i + 2)->Uint32Value();
+      Local<Value> node_field;
+      if (Nan::Get(js_node, i + 2).ToLocal(&node_field)) {
+        node.context[i] = Nan::To<uint32_t>(node_field).FromMaybe(0);
+      }
     }
 
     ts_node_edit(&node, &edit);
 
     for (unsigned i = 0; i < 4; i++) {
-      js_node->Set(i + 2, Nan::New(node.context[i]));
+      Nan::Set(js_node, i + 2, Nan::New(node.context[i]));
     }
   }
 
@@ -135,7 +139,7 @@ void Tree::GetChangedRanges(const Nan::FunctionCallbackInfo<Value> &info) {
 
   Local<Array> result = Nan::New<Array>();
   for (size_t i = 0; i < range_count; i++) {
-    result->Set(i, RangeToJS(ranges[i]));
+    Nan::Set(result, i, RangeToJS(ranges[i]));
   }
 
   info.GetReturnValue().Set(result);
@@ -197,7 +201,6 @@ void Tree::PrintDotGraph(const Nan::FunctionCallbackInfo<Value> &info) {
 
 static void FinalizeNode(const v8::WeakCallbackInfo<Tree::NodeCacheEntry> &info) {
   Tree::NodeCacheEntry *cache_entry = info.GetParameter();
-  assert(cache_entry->node.IsNearDeath());
   assert(!cache_entry->node.IsEmpty());
   cache_entry->node.Reset();
   if (cache_entry->tree) {
@@ -211,16 +214,18 @@ void Tree::CacheNode(const Nan::FunctionCallbackInfo<Value> &info) {
   Tree *tree = ObjectWrap::Unwrap<Tree>(info.This());
   Local<Object> js_node = Local<Object>::Cast(info[0]);
 
+  Local<Value> js_node_field1, js_node_field2;
+  if (!Nan::Get(js_node, 0).ToLocal(&js_node_field1)) return;
+  if (!Nan::Get(js_node, 1).ToLocal(&js_node_field2)) return;
   uint32_t key_parts[2] = {
-    js_node->Get(0)->Uint32Value(),
-    js_node->Get(1)->Uint32Value(),
+    Nan::To<uint32_t>(js_node_field1).FromMaybe(0),
+    Nan::To<uint32_t>(js_node_field2).FromMaybe(0)
   };
   const void *key = UnmarshalNodeId(key_parts);
 
   auto cache_entry = new NodeCacheEntry{tree, key, {}};
   cache_entry->node.Reset(info.GetIsolate(), js_node);
   cache_entry->node.SetWeak(cache_entry, &FinalizeNode, Nan::WeakCallbackType::kParameter);
-  cache_entry->node.MarkIndependent();
 
   assert(!tree->cached_nodes_.count(key));
 

--- a/src/tree_cursor.cc
+++ b/src/tree_cursor.cc
@@ -49,8 +49,8 @@ void TreeCursor::Init(v8::Local<v8::Object> exports) {
     Nan::SetPrototypeMethod(tpl, methods[i].name, methods[i].callback);
   }
 
-  Local<Function> constructor_local = tpl->GetFunction();
-  exports->Set(class_name, constructor_local);
+  Local<Function> constructor_local = Nan::GetFunction(tpl).ToLocalChecked();
+  Nan::Set(exports, class_name, constructor_local);
   constructor.Reset(Nan::Persistent<Function>(constructor_local));
 }
 
@@ -87,10 +87,12 @@ void TreeCursor::GotoFirstChild(const Nan::FunctionCallbackInfo<Value> &info) {
 
 void TreeCursor::GotoFirstChildForIndex(const Nan::FunctionCallbackInfo<Value> &info) {
   TreeCursor *cursor = Nan::ObjectWrap::Unwrap<TreeCursor>(info.This());
-  if (!info[0]->IsUint32()) {
+  auto maybe_index = Nan::To<uint32_t>(info[0]);
+  if (maybe_index.IsNothing()) {
     Nan::ThrowTypeError("Argument must be an integer");
+    return;
   }
-  uint32_t goal_byte = info[0]->Uint32Value() * 2;
+  uint32_t goal_byte = maybe_index.FromJust() * 2;
   int64_t child_index = ts_tree_cursor_goto_first_child_for_byte(&cursor->cursor_, goal_byte);
   if (child_index < 0) {
     info.GetReturnValue().Set(Nan::Null());
@@ -120,7 +122,7 @@ void TreeCursor::EndPosition(const Nan::FunctionCallbackInfo<Value> &info) {
 void TreeCursor::CurrentNode(const Nan::FunctionCallbackInfo<Value> &info) {
   TreeCursor *cursor = Nan::ObjectWrap::Unwrap<TreeCursor>(info.This());
   Local<String> key = Nan::New<String>("tree").ToLocalChecked();
-  const Tree *tree = Tree::UnwrapTree(info.This()->Get(key));
+  const Tree *tree = Tree::UnwrapTree(Nan::Get(info.This(), key).ToLocalChecked());
   TSNode node = ts_tree_cursor_current_node(&cursor->cursor_);
   node_methods::MarshalNode(info, tree, node);
 }
@@ -128,7 +130,7 @@ void TreeCursor::CurrentNode(const Nan::FunctionCallbackInfo<Value> &info) {
 void TreeCursor::Reset(const Nan::FunctionCallbackInfo<Value> &info) {
   TreeCursor *cursor = Nan::ObjectWrap::Unwrap<TreeCursor>(info.This());
   Local<String> key = Nan::New<String>("tree").ToLocalChecked();
-  const Tree *tree = Tree::UnwrapTree(info.This()->Get(key));
+  const Tree *tree = Tree::UnwrapTree(Nan::Get(info.This(), key).ToLocalChecked());
   TSNode node = node_methods::UnmarshalNode(tree);
   ts_tree_cursor_reset(&cursor->cursor_, node);
 }

--- a/test/node_test.js
+++ b/test/node_test.js
@@ -210,7 +210,7 @@ describe("Node", () => {
   describe(".descendantForPosition(min, max)", () => {
     it("returns the smallest node that spans the given range", () => {
       const tree = parser.parse("x10 + 1000");
-      const sumNode = tree.rootNode.firstChild;
+      const sumNode = tree.rootNode.firstChild.firstChild;
 
       assert.equal(
         "identifier",
@@ -227,11 +227,11 @@ describe("Node", () => {
 
       assert.throws(() => {
         sumNode.descendantForPosition(1, {});
-      }, /Point.row must be a number/);
+      }, /Point must be a {row, column} object/);
 
       assert.throws(() => {
         sumNode.descendantForPosition();
-      }, /Point must be a .* object/);
+      }, /Point must be a {row, column} object/);
     });
   });
 

--- a/test/parser_test.js
+++ b/test/parser_test.js
@@ -100,7 +100,7 @@ describe("Parser", () => {
 
     it("reads from the given input", () => {
       const parts = ["first", "_", "second", "_", "third"];
-      const tree = parser.parse(() => parts.shift());
+      const tree = parser.parse((index) => parts.shift());
       assert.equal(tree.rootNode.toString(), "(program (expression_statement (identifier)))");
     });
 

